pkg/container-disk/container-disk.go

```diff
@@ -53,15 +53,27 @@ func GetVolumeMountDirOnGuest(vmi *v1.VirtualMachineInstance) string {
 }
 
 func GetVolumeMountDirOnHost(vmi *v1.VirtualMachineInstance) (string, bool, error) {
+	basepath := ""
+	foundEntries := 0
+	foundBasepath := ""
 	for podUID, _ := range vmi.Status.ActivePods {
-		basepath := fmt.Sprintf("%s/%s/volumes/kubernetes.io~empty-dir/container-disks", podsBaseDir, string(podUID))
+		basepath = fmt.Sprintf("%s/%s/volumes/kubernetes.io~empty-dir/container-disks", podsBaseDir, string(podUID))
 		exists, err := diskutils.FileExists(basepath)
 		if err != nil {
 			return "", false, err
 		} else if exists {
-			return basepath, true, nil
+			foundEntries++
+			foundBasepath = basepath
 		}
 	}
+
+	if foundEntries == 1 {
+		return foundBasepath, true, nil
+	} else if foundEntries > 1 {
+		// Don't mount until outdated pod environments are removed
+		// otherwise we might stomp on a previous cleanup
+		return "", false, fmt.Errorf("Found multiple pods active for vmi %s/%s. Waiting on outdated pod directories to be removed", vmi.Namespace, vmi.Name)
```

Comment:
Should we in this case unmount the mounts from the dead pod explicitly?

Reply:
I've gone back and forth on this.

My thought process is that I don't want to make this mount/unmount logic too intelligent. The way it is now, this logic just gates itself when it detects an inconsistency in the environment, and waits for the eventual consistent nature of the vm controller and kubelet to eventually do the right thing. If we make this function any smarter, then it has to make a decision about what the "right" and "wrong" pods are when it comes to mounting. If we just make it gate the logic on an inconsistency, then we are deferring to the system to correct the issue in time.

99% of the time, this gating logic is hit because of the kubelet backoff during container teardown that occurs when we don't umount the old container fast enough. The controller will umount the old container due to the ghost file eventually, but fully terminating the container after that is dependent on when the kubelet backoff logic fires... so it looks like there are two active pod directories because the kubelet hasn't completed the teardown yet.